File: Snooper/src/main/java/com/prateekj/snooper/infra/BackgroundTaskExecutor.java
package com.prateekj.snooper.infra;

import android.app.Activity;
import android.os.AsyncTask;

public class BackgroundTaskExecutor {
  private Activity activity;

  public BackgroundTaskExecutor(Activity activity) {
    this.activity = activity;
  }

  public <E> void execute(final BackgroundTask<E> backgroundTask) {
    AsyncTask.execute(new Runnable() {
      @Override
      public void run() {
        E result = backgroundTask.onExecute();
        sendResult(result, backgroundTask);
      }
    });
  }

  public <E> void sendResult(final E result, final BackgroundTask<E> backgroundTask) {
    this.activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        backgroundTask.onResult(result);
      }
    });
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/infra/BackgroundTaskExecutor.kt
package com.prateekj.snooper.infra

import android.app.Activity
import android.os.AsyncTask

class BackgroundTaskExecutor(private val activity: Activity) {

  fun <E> execute(backgroundTask: BackgroundTask<E>) {
    AsyncTask.execute {
      val result = backgroundTask.onExecute()
      sendResult(result, backgroundTask)
    }
  }

  fun <E> sendResult(result: E, backgroundTask: BackgroundTask<E>) {
    this.activity.runOnUiThread { backgroundTask.onResult(result) }
  }
}
